### DISCUSSION TITLE
Efficient way to parse shell command output (stdout) with nushell

### DISCUSSION THREAD

**Discussion by metacoma** [nushell/nushell#7430] [Q&A] [answered]
Hey folks, i do research/experiments in UX.

https://user-images.githubusercontent.com/5146707/206879310-0dcc179d-6845-46af-b869-c1cd6d86bb8f.mp4

I try to find the efficient way to parse the output of shell commands into some structured data and then transform/template this data to the groovy code.

I have tried  it in several ways:

1. hard way with shell/awk/sed
https://github.com/metacoma/mindwm/blob/main/mindmap-shell/shell2mindmap.sh
https://github.com/metacoma/mindwm/blob/main/adapters/tmux.sh

2. Another try with difficult jinja2 template
https://github.com/metacoma/mindwm/blob/master/templates/mindwm/context_match.py

And so on.

Recently, i started play with `nushell` and at this moment i have this ugly script
https://github.com/metacoma/mindwm/blob/master/files/nushell/nushell_parser.sh
Ugly script accepts the command output as stdin, then the output parsed by nushell inside docker container and finally, the result (output) from nushell parsed by awk

The text transformation be like:
source data
```
CONTAINER ID   IMAGE                         COMMAND                  CREATED        STATUS                  PORTS                                                                                    NAMES
6e0483ece958   geoffh1977/chrony             "tini -- /usr/local/…"   3 months ago   Up 2 months             0.0.0.0:123->123/udp, :::123->123/udp                                                    chrony
5ff249236853   ubuntu/squid:5.2-22.04_beta   "entrypoint.sh -f /e…"   3 months ago   Up 2 months             0.0.0.0:3128->3128/tcp, :::3128->3128/tcp                                                squid-container
1dcd4580bf93   balabit/syslog-ng:latest      "/usr/sbin/syslog-ng…"   3 months ago   Up 2 months (healthy)   0.0.0.0:514->514/udp, :::514->514/udp, 0.0.0.0:601->601/tcp, :::601->601/tcp, 6514/tcp   syslog-ng
```

After nushell, this text will transform into this pretty table
`nushell nu --stdin -c 'detect columns | table -w 1000'`
```
╭───┬──────────────┬──────────────┬─────────────────────────────┬────────────────────────┬─────────┬────────┬─────────────────────────┬─────────────────╮
│ # │  CONTAINER   │      ID      │            IMAGE            │        COMMAND         │ CREATED │ STATUS │          PORTS          │      NAMES      │
├───┼──────────────┼──────────────┼─────────────────────────────┼────────────────────────┼─────────┼────────┼─────────────────────────┼─────────────────┤
│ 0 │ 6e0483ece958 │ 6e0483ece958 │ geoffh1977/chrony           │ "tini -- /usr/local/…" │ months  │ 2      │ 0.0.0.0:123->123/udp,   │ chrony          │
│ 1 │ 5ff249236853 │ 5ff249236853 │ ubuntu/squid:5.2-22.04_beta │ "entrypoint.sh -f /e…" │ months  │ 2      │ 0.0.0.0:3128->3128/tcp, │ squid-container │
│ 2 │ 1dcd4580bf93 │ 1dcd4580bf93 │ balabit/syslog-ng:latest    │ "/usr/sbin/syslog-ng…" │ months  │ 2      │ 0.0.0.0:514->514/udp,   │ syslog-ng       │
╰───┴──────────────┴──────────────┴─────────────────────────────┴────────────────────────┴─────────┴────────┴─────────────────────────┴─────────────────╯

```
and then, after the awk script, i got the following groovy code:
```groovy
chrony = node.createChild("chrony")
chrony["#"] = "0"
chrony["CONTAINER"] = "6e0483ece958"
chrony["ID"] = "6e0483ece958"
chrony["IMAGE"] = "geoffh1977/chrony"
chrony["COMMAND"] = "tini -- /usr/local/..."
chrony["CREATED"] = "months"
chrony["STATUS"] = "2"
chrony["PORTS"] = "0.0.0.0:123->123/udp,"
chrony["NAMES"] = "chrony"
squid_container = node.createChild("squid_container")
squid_container["#"] = "1"
squid_container["CONTAINER"] = "5ff249236853"
squid_container["ID"] = "5ff249236853"
squid_container["IMAGE"] = "ubuntu/squid:5.2-22.04_beta"
squid_container["COMMAND"] = "entrypoint.sh -f /e..."
squid_container["CREATED"] = "months"
squid_container["STATUS"] = "2"
squid_container["PORTS"] = "0.0.0.0:3128->3128/tcp,"
squid_container["NAMES"] = "squid-container"
syslog_ng = node.createChild("syslog_ng")
syslog_ng["#"] = "2"
syslog_ng["CONTAINER"] = "1dcd4580bf93"
syslog_ng["ID"] = "1dcd4580bf93"
syslog_ng["IMAGE"] = "balabit/syslog-ng:latest"
syslog_ng["COMMAND"] = "/usr/sbin/syslog-ng..."
syslog_ng["CREATED"] = "months"
syslog_ng["STATUS"] = "2"
syslog_ng["PORTS"] = "0.0.0.0:514->514/udp,"
syslog_ng["NAMES"] = "syslog-ng"
```
You can see how it work in the attachment video.

Also, i want to pay attention, that i ran commands on the remote host where `nushell` is not installed. (it also works, inside docker/k8s containers and, for example, on the remote network switch). But i have ability to pass the output from the command that in ran on the remote hosts, to the 'local' `nushell` instance.

So, my question, can `nushell` easier my life in text transformations? I guess, yes.
Please feel free to advise

**Comment by fdncred** [comment]
Interesting experiment. Another command to use for this nushell --stdin stuff is `from ssv`. There are a few interesting parameters you can see in `help from ssv`.

**Comment by kubouch** [comment, marked as answer]
Also, the `parse` command is very useful. I use it often when capturing some stdout output. Especially the `parse -r` is very powerful. It might be a bit hard to learn but the examples in `parse -h` should help enough to get started.

**Comment by metacoma** [comment]
hey

Thank you for your suggestions. 
Looks like I found a easier way how to parse text.

I will use logstash https://github.com/elastic/logstash for this
it already contains the parsers/filters I need

Just take a look into this small logstash pipeline
https://github.com/metacoma/mindwm/blob/main/files/logstash/pipeline/kubernetes.conf

And the result in UI

https://user-images.githubusercontent.com/5146707/207229948-f1c6cf60-1bcc-4104-96b7-2095247855db.mp4
